### PR TITLE
Miscellaneous fixes

### PR DESCRIPTION
**Pull request by DonJayamanne**
Some of the debt work from another PR thats a WIP

### CODE REVIEW

File: src/interactive-window/interactiveWindow.ts [hunk 1]
```diff
@@ -422,26 +422,7 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
         message = message.split('\n').join('  \n');
         this.updateSysInfoMessage(message, true, cellPromise);
     }
-    private registerControllerChangeListener(controller: IVSCodeNotebookController) {
```
Comment (DonJayamanne) [inline src/interactive-window/interactiveWindow.ts:425]: This is pointless we already dispose the kernel in another place line 435

Comment (DonJayamanne) [reply src/interactive-window/interactiveWindow.ts:425]: Changes from https://github.com/microsoft/vscode-jupyter/pull/10573

File: src/standalone/api/kernelApi.ts [hunk 2]
```diff
@@ -247,13 +247,15 @@ class JupyterKernelService implements IExportedKernelService {
...
-            const readWriteConnection = connection as IKernelKernelConnectionMetadata;
             // By not forcing the cast, we ensure the types are compatible.
             // All we're doing is ensuring the readonly version of one type is compatible with the other.
-            // Also, we must return a readonly version of the type (to prevent anyone from stuffing this).
+            // Also, we must return a readonly version of the type (to prevent 3rd party extensions from stuffing this).
             // Else it breaks the Jupyter extension
             // We recast to KernelConnectionMetadata as this has already define its properties as readonly.
-            const translatedConnection = Object.freeze(readWriteConnection) as KernelConnectionMetadata;
+
+            const translatedConnection = Object.freeze(
```
Comment (DonJayamanne) [inline src/standalone/api/kernelApi.ts:256]: Code copied from another PR
Better way of freezing the kernel connection

Changes from https://github.com/microsoft/vscode-jupyter/pull/10573

File: src/test/datascience/interactiveWindow.vscode.common.test.ts [hunk 3]
```diff
@@ -112,6 +112,15 @@ suite(`Interactive window execution`, async function () {
         assert.equal(actualSource, source, `Executed cell has unexpected source code`);
         await waitForExecutionCompletedSuccessfully(secondCell!);
         await waitForTextOutput(secondCell!, '42');
+
+        // If we're connected to a remote kernel,
+        // then ensure the kernel points to a live connection.
+        if (IS_REMOTE_NATIVE_TEST()) {
+            assert.strictEqual(
```
Comment (DonJayamanne) [inline src/test/datascience/interactiveWindow.vscode.common.test.ts:119]: Changes from https://github.com/microsoft/vscode-jupyter/pull/10573

File: src/test/datascience/widgets/commUtils.ts [hunk 4]
```diff
@@ -3,15 +3,13 @@
...
+import { traceInfo, traceInfoIfCI } from '../../../platform/logging';
+import { IDisposable } from '../../../platform/common/types';
 import { createDeferred } from '../../../platform/common/utils/async';
-import { IServiceContainer } from '../../../platform/ioc/types';
 import { noop } from '../../core';
 import * as colors from 'colors';
 
-export function initializeWidgetComms(serviceContainer: IServiceContainer): Utils {
-    const disposables = serviceContainer.get<IDisposableRegistry>(IDisposableRegistry);
+export function initializeWidgetComms(disposables: IDisposable[]): Utils {
```
Comment (DonJayamanne) [inline src/test/datascience/widgets/commUtils.ts:12]: Changes from https://github.com/microsoft/vscode-jupyter/pull/10573

File: src/webviews/webview-side/ipywidgets/kernel/index.ts [hunk 5]
```diff
@@ -131,9 +130,13 @@ function renderIPyWidget(
     logger: (message: string, category?: 'info' | 'error') => void
 ) {
     logger(`Rendering IPyWidget ${outputId} with model ${model.model_id}`);
-    if (renderedWidgets.has(outputId)) {
+    if (renderedWidgets.has(outputId) && renderedWidgets.get(outputId)?.container === container) {
         return logger('already rendering');
     }
+    if (renderedWidgets.has(outputId)) {
+        logger('Widget was already rendering for another container, dispose that widget so we can re-render it');
```
Comment (DonJayamanne) [inline src/webviews/webview-side/ipywidgets/kernel/index.ts:137]: Changes from https://github.com/microsoft/vscode-jupyter/pull/10573

File: src/interactive-window/interactiveWindow.ts [hunk 6]
```diff
@@ -422,26 +422,7 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
...
-        const controllerChangeListener = controller.controller.onDidChangeSelectedNotebooks(
-            (selectedEvent: { notebook: NotebookDocument; selected: boolean }) => {
-                // Controller was deselected for this InteractiveWindow's NotebookDocument
-                if (selectedEvent.selected === false && selectedEvent.notebook === this.notebookEditor.notebook) {
-                    controllerChangeListener.dispose();
-                    this.disconnectKernel();
```
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/interactive-window/interactiveWindow.ts:431]: There's no need, if you check list 454, when we change the controller, we disconnect the controller.

**Review by rchiodo — APPROVED** [review]

**Pull request merged**